c7n/resources/s3.py

```diff
@@ -799,6 +800,106 @@ def __call__(self, b):
         return True
 
 
+class BucketNotificationFilter(ValueFilter):
+    """Base class for bucket notification filters"""
+
+    def process(self, buckets, event=None):
+        return super(BucketNotificationFilter, self).process(buckets, event)
+
+    def __call__(self, bucket):
+
+        found_configuration = False
+        for configuration in bucket.get('Notification', {}).get(self.configurations, []):
+            if self.match(configuration):
+                set_annotation(
+                    bucket,
+                    'c7n:MatchedNotificationConfigurationIds',
+                    configuration['Id'])
+                found_configuration = True
+        return found_configuration
+
+
+@filters.register('lambda-function-notification')
+class LambdaFunctionNotificationFilter(BucketNotificationFilter):
+    """Filter based on bucket lambda notification configuration.
```

Comment:
how about collapsing these four classes and three filters into a single one. ie 

```
type: notification 
kind: lambda|sns|sqs```

Reply:
reasoning is keeping the filter set fairly orthogonal and to expose minor differentiation via parameterization rather than expansion of the vocabulary for undifferentiated capability deltas.